### DISCUSSION TITLE
Rate-limiting

### DISCUSSION THREAD

**Discussion by tyteen4a03** [next-safe-action/next-safe-action#272] [Q&A]
Is there a recommended way of implementing rate-limiting for server actions (e.g. to prevent brute-force login attacks)?

**Comment by JBotwina** [comment]
@tyteen4a03 What about adding an [upstash rate limiter](https://upstash.com/docs/redis/sdks/ratelimit-ts/gettingstarted) to the middleware?

**Reply by tyteen4a03** [reply]
This package seems to be tied to Upstash Redis / Vercel KV - was looking for something that works with self-hosted Redis or Cloudflare KV.
